build.assets/tooling/cmd/build-apt-repos/apt_repo_tool.go

```diff
@@ -0,0 +1,210 @@
+	}
+
+	err = art.publishRepos()
+	if err != nil {
+		return trace.Wrap(err, "failed to publish repos")
+	}
+
+	err = art.s3Manager.UploadBuiltRepo(filepath.Join(art.aptly.rootDir, "public"))
+	if err != nil {
+		return trace.Wrap(err, "failed to sync changes to S3 bucket")
+	}
+
+	logrus.Infof("APT repo build process completed in %s", time.Since(start).Round(time.Millisecond))
+	return nil
+}
+
+func (art *AptRepoTool) publishRepos() error {
+	// Pull in all Aptly repos, not just the latest ones to ensure they all get built into APT repos correctly
+	repos, err := art.aptly.GetAllRepos()
+	if err != nil {
+		return trace.Wrap(err, "failed to get all Aptly repos")
+	}
+
+	// Build a map keyed by os info with value of all repos that support the os in the key
+	// This will be used to structure the publish command
+	logrus.Debugf("Categorizing repos according to OS info: %v", RepoNames(repos))
+	categorizedRepos := make(map[string][]*Repo)
+	for _, r := range repos {
+		if osRepos, ok := categorizedRepos[r.OSInfo()]; ok {
+			categorizedRepos[r.OSInfo()] = append(osRepos, r)
+		} else {
+			categorizedRepos[r.OSInfo()] = []*Repo{r}
+		}
+	}
+	logrus.Debugf("Categorized repos: %v", categorizedRepos)
+
+	for osInfo, osRepoList := range categorizedRepos {
+		if len(osRepoList) < 1 {
+			continue
+		}
+
+		err := art.aptly.PublishRepos(osRepoList, osRepoList[0].os, osRepoList[0].osVersion)
+		if err != nil {
+			return trace.Wrap(err, "failed to publish for os %q", osInfo)
+		}
+	}
+
+	return nil
+}
+
+func (art *AptRepoTool) recreateExistingRepos(localPublishedPath string) ([]*Repo, error) {
+	logrus.Infoln("Recreating previously published repos...")
+	createdRepos, err := art.aptly.CreateReposFromPublishedPath(localPublishedPath)
+	if err != nil {
+		return nil, trace.Wrap(err, "failed to recreate existing repos")
+	}
+
+	for _, repo := range createdRepos {
+		err := art.aptly.ImportDebsFromExistingRepo(repo)
+		if err != nil {
+			return nil, trace.Wrap(err, "failed to import debs from existing repo %q", repo.Name())
+		}
+	}
+
+	logrus.Infof("Recreated and imported pre-existing artifacts for %d repos", len(createdRepos))
+	return createdRepos, nil
+}
+
+func (art *AptRepoTool) getArtifactRepos() ([]*Repo, error) {
+	logrus.Infoln("Creating or getting Aptly repos for artifact requirements...")
+
+	artifactRepos, err := art.aptly.CreateReposFromArtifactRequirements(art.supportedOSs,
+		art.config.releaseChannel, semver.Major(art.config.artifactVersion))
+	if err != nil {
+		return nil, trace.Wrap(err, "failed to create or get repos from artifact requirements")
+	}
+
+	logrus.Infof("Created or got %d artifact Aptly repos", len(artifactRepos))
+	return artifactRepos, nil
+}
+
+func (art *AptRepoTool) importNewDebs(repos []*Repo) error {
+	logrus.Debugf("Importing new debs into %d repos: %q", len(repos), strings.Join(RepoNames(repos), "\", \""))
+	err := filepath.WalkDir(art.config.artifactPath,
+		func(debPath string, d fs.DirEntry, err error) error {
```

Comment:
Nit: For clarity, I might declare this `WalkFunc` ahead of time instead of inline.  As I was reviewing, I found by the time I got to the bottom of the anonomous function, I'd lost the `importNewDebs` context.